# SpatialOmicsOverlay Workshop
## Bioconductor Smorgasbord 2023

Welcome to the workshop for NanoString's SpatialOmicsOverlay R package. 

## Instructor names
* Maddy Griswold<sup>1</sup>
* Megan Vandenberg<sup>1</sup>
* Stephanie Zimmerman<sup>1</sup>

<sup>1</sup>NanoString Technologies, Inc.

## Workshop Description
 This workshop will introduce users to the NanoString R package SpatialOmicsOverlay. This package is designed for use in interpretation and presentation of the multi-form data generated by NanoString's [GeoMx <sup>&reg;</sup> Digital Spatial Profiler](https://nanostring.com/products/geomx-digital-spatial-profiler/geomx-dsp-overview/) spatial biology platform. The GeoMx DSP produces both rich imaging and genomics data. Integrating these data types facilitates a deep understanding of the profiled tissue. The SpatialOmicsOverlay package integrates both of these data types, thereby by maintaining the relationship between underlying tissue morphology and resultant gene expression. 

Specifically, SpatialOmicsOverlay was developed to visualize and analyze the free-handed nature of Region of Interest (ROI) selection in a GeoMx experiment, as well as the immunofluorescence-guided segmentation process. The overlay from the instrument is recreated in the R environment, which allows for plotting overlays with data like ROI type or gene expression. The package provides a convenient workflow for users to generate customized, sharable visualizations.

The *Introduction to SpatialOmicsOverlay* vignette demonstrates how to use OME-TIFF files, which are exported from the GeoMx platform. Participants will learn how to interact with this file type and generate informative plots over images. This vignette utilizes data from our [Spatial Organ Atlas](https://nanostring.com/products/geomx-digital-spatial-profiler/spatial-organ-atlas/). The Spatial Organ Atlas is a freely-accesible resource of whole transcriptome spatial profiles of functional components of tissues from human and mouse generated using our [Whole Transcriptome Atlas RNA assay](https://doi.org/10.1101/2021.09.29.462442). In particular, vignette users will be analyzing data from the mouse brain. This content is similar to the vignette available with the package upon installation.

Mouse brain with ROIs only - see ROIs in space:

![](https://github.com/Nanostring-Biostats/SOOBiocWorkshop/blob/devel/images/fullBrainNoImage.png?raw=true)

Mouse brain with ROIs and immunofluoresence image - see substructures samples:

![](https://github.com/Nanostring-Biostats/SOOBiocWorkshop/blob/devel/images/fullBrainwImage.png?raw=true)

Mouse brain with Calm1 expression - see relationship between tissue structure and gene expression:

![](https://github.com/Nanostring-Biostats/SOOBiocWorkshop/blob/devel/images/fullBrainwImageAndCalm1.png?raw=true)

## Pre-requisites
* Basic knowledge of R programming
* Docker installed and ready to run
* Space on machine hosting Docker image to download a 13 GB demo data file.
* Optional: familiarity with NanoString's DSP platform and data type. See *Resources* section.


## Download the Docker image

```sh
docker run -e PASSWORD=<yourchosenpassword> -p 8787:8787 ghcr.io/nanostring-biostats/soobiocworkshop
```
Once running, navigate to http://localhost:8787/ and then login with `rstudio`:`<yourchosenpassword>`. 


*NOTE*: Running docker that uses the password in plain text like above exposes the password to others 
in a multi-user system (like a shared workstation or compute node). In practice, consider using an environment 
variable instead of plain text to pass along passwords and other secrets in docker command lines. 

If, upon running the above command, you receive the following error:
```sh
Unable to find image 'ghcr.io/nanostring-biostats/soobiocworkshop:latest' locally
docker: Error response from daemon: Head "https://ghcr.io/v2/nanostring-biostats/soobiocworkshop/manifests/latest": unauthorized.
See 'docker run --help'.
```
You may need to log into the GitHub Container Registry (GHCR) using:
```sh
docker login ghcr.io
```
Your usename will be your GitHub username and your password should be a classic-style personal access token. See the GitHub page ["Working with the Container registry"](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry).


## Resources
* Explore the [workshop homepage](https://nanostring-biostats.github.io/SOOBiocWorkshop/index.html).
* Learn more about the [GeoMx DSP platform](https://nanostring.com/products/geomx-digital-spatial-profiler/geomx-dsp-overview/).
* Access [SpatialOmicsOverlay](https://bioconductor.org/packages/release/bioc/html/SpatialOmicsOverlay.html) on Bioconductor.
* Explore the [Spatial Organ Atlas](https://nanostring.com/products/geomx-digital-spatial-profiler/spatial-organ-atlas/) data on NanoString's website.
* View [source code](https://github.com/Nanostring-Biostats/SOOBiocWorkshop) for this workshop.